docs/src/tutorial-nondeterministic-variables.md

```diff
@@ -1,95 +1,103 @@
+There are no arguments provided to `kani::any()` to indicate an upper bound.
 
-```rust
-{{#include tutorial/arbitrary-variables/src/inventory.rs:unsafe_update}}
-```
+You can use `kani::any()` for `[T; N]` (if implemented for `T`).
+But this the type of an exactly-sized array.
+Again, `kani::any()` does not have an implementation for an unsized slice (like `[T]`).
 
-We commented out the assertion that the underlying `u32` variable cannot be `0`, since this no longer holds.
-The verification will now fail showing that the `inventory.get(&id).unwrap()` method call can panic.
+This does not mean you cannot have a nondeterministic vector.
+It just means you have to construct one.
+Our example proof harness above constructs a nondeterministic `Inventory` of size `1`, simply by starting with the empty `Inventory` and inserting a nondeterministic entry.
 
-This is an interesting issue that emerges from how `rustc` optimizes the memory layout of `Option<NonZeroU32>`.
-The compiler is able to represent `Option<NonZeroU32>` using `32` bits by using the value `0` to represent `None`.
+## Custom nondeterministic types
 
-You can try it out by running the example under [`arbitrary-variables`](https://github.com/model-checking/kani/tree/main/docs/src/tutorial/arbitrary-variables/):
+When you need nondeterministic variables of types that `kani::any()` cannot construct, our current recommendation is to simply write a function.
 
-```
-cargo kani --harness unsafe_update
+To generate a nondeterministic struct, you would just generate nondeterministic values for each of its fields.
+For complex data structures, you can start with an empty one and add a (bounded) nondeterministic number of entries.
+For an enum, you can make use of a simple trick:
+
+```rust
+{{#include tutorial/arbitrary-variables/src/rating.rs:rating_invariant}}
 ```
 
-## Safe nondeterministic variables for custom types
+All we're doing here is making use of a nondeterministic integer to decide which variant of `Rating` to return.
 
-Now you would like to add a new structure to your library that allow users to represent a review rating, which can go from 0 to 5 stars.
-Let's say you add the following implementation:
+> **NOTE**: If we thought of this code as generating a random value, this function looks sub-optimal.
+> We'd overwhelming generate a `Three` because it's matching "all other integers besides 1 and 2."
+> But Kani just see 3 meaningful possibilities, each of which is not treated any differently from each other.
+> The "proportion" of integers does not matter.
 
-```rust
-{{#include tutorial/arbitrary-variables/src/rating.rs:rating_struct}}
-```
 
-The easiest way to allow users to create nondeterministic variables of the Rating type which represents values from 0-5 stars is by implementing the `kani::Invariant` trait.
+## Exercise
```

Comment:
I like this. Should we provide an answer for this? Maybe refer to a file that they can open at the end.